File: csharp/Betfair.ESAClient/Betfair.ESAClient/Cache/OrderCache.cs
using Betfair.ESAClient.Protocol;
using Betfair.ESASwagger.Model;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Betfair.ESAClient.Cache
{
    public class OrderCache
    {
        private readonly ConcurrentDictionary<string, OrderMarket> _markets = new ConcurrentDictionary<string, OrderMarket>();

        public OrderCache()
        {
            IsOrderMarketRemovedOnClose = true;
        }

        public void OnOrderChange(ChangeMessage<OrderMarketChange> changeMessage)
        {
            if (changeMessage.IsStartOfNewSubscription)
            {
                //clear cache
                _markets.Clear();
            }
            if(changeMessage.Items != null)
            {
                //lazy build events
                List<OrderMarketChangedEventArgs> batch = BatchOrderMarketsChanged == null ? null : new List<OrderMarketChangedEventArgs>(changeMessage.Items.Count);

                foreach (OrderMarketChange marketChange in changeMessage.Items)
                {
                    bool isImage = marketChange.FullImage == true;
                    if (isImage) {
                        // Clear market from cache if it is being re-imaged
                        OrderMarket removed;
                        _markets.TryRemove(marketChange.Id, out removed);
                    }

                    OrderMarket market = OnOrderMarketChange(marketChange);

                    if(IsOrderMarketRemovedOnClose && market.IsClosed)
                    {
                        //remove on close
                        OrderMarket removed;
                        _markets.TryRemove(market.MarketId, out removed);
                    }
                    //lazy build events
                    if (batch != null || OrderMarketChanged != null)
                    {
                        OrderMarketChangedEventArgs arg = new OrderMarketChangedEventArgs() { Change = marketChange, OrderMarket = market };
                        if (OrderMarketChanged != null)
                        {
                            DispatchOrderMarketChanged(arg);
                        }
                        if (batch != null)
                        {
                            batch.Add(arg);
                        }
                    }
                }
                if (batch != null)
                {
                    DispatchBatchOrderMarketsChanged(new BatchOrderMarketsChangedEventArgs() { Changes = batch });
                }
            }
        }

        private OrderMarket OnOrderMarketChange(OrderMarketChange marketChange)
        {
            OrderMarket market = _markets.GetOrAdd(marketChange.Id, id => new OrderMarket(this, id));
            market.OnOrderMarketChange(marketChange);
            return market;
        }

        private void DispatchOrderMarketChanged(OrderMarketChangedEventArgs args)
        {
            try
            {
                OrderMarketChanged.Invoke(this, args);
            }
            catch (Exception e)
            {
                Trace.TraceError("Error dispatching event: {0}", e);
            }
        }

        private void DispatchBatchOrderMarketsChanged(BatchOrderMarketsChangedEventArgs args)
        {
            try
            {
                BatchOrderMarketsChanged.Invoke(this, args);
            }
            catch (Exception e)
            {
                Trace.TraceError("Error dispatching event: {0}", e);
            }
        }

        /// <summary>
        /// Wether order markets are automatically removed on close
        /// (default is true)
        /// </summary>
        public bool IsOrderMarketRemovedOnClose { get; set; }

        /// <summary>
        /// Event for each order market change
        /// </summary>
        public event OrderMarketChangedEventHandler OrderMarketChanged;

        /// <summary>
        /// Event for each batch of order market changes
        /// (note to be truly atomic you will want to set to merge segments
        /// otherwise an event could be segmented)
        /// </summary>
        public event BatchOrderMarketsChangedEventHandler BatchOrderMarketsChanged;

        /// <summary>
        /// Queries by market id - the result is invariant for the 
        /// lifetime of the market.
        /// </summary>
        /// <param name="marketid"></param>
        /// <returns></returns>
        public OrderMarket this[string marketId]
        {
            get
            {
                return _markets[marketId];
            }
        }

        /// <summary>
        /// All the cached markets
        /// </summary>
        public IEnumerable<OrderMarket> Markets
        {
            get
            {
                return _markets.Values;
            }
        }

        /// <summary>
        /// Market count
        /// </summary>
        public int Count
        {
            get
            {
                return _markets.Count;
            }
        }
    }

    public delegate void OrderMarketChangedEventHandler(object sender, OrderMarketChangedEventArgs e);

    public delegate void BatchOrderMarketsChangedEventHandler(object sender, BatchOrderMarketsChangedEventArgs e);

    public class OrderMarketChangedEventArgs : EventArgs
    {
        /// <summary>
        /// The raw change message that was just applied
        /// </summary>
        public OrderMarketChange Change {get; internal set;}

        /// <summary>
        /// The order market changed - this is reference invariant
        /// </summary>
        public OrderMarket OrderMarket { get; internal set; }

        /// <summary>
        /// Takes or returns an existing immutable snap of the market.
        /// </summary>
        public OrderMarketSnap Snap
        {
            get
            {
                return OrderMarket.Snap;
            }
        }
    }

    public class BatchOrderMarketsChangedEventArgs : EventArgs
    {
        public IList<OrderMarketChangedEventArgs> Changes { get; internal set; }
    }
}

File: java/client/src/main/java/com/betfair/esa/client/cache/order/OrderCache.java
package com.betfair.esa.client.cache.order;

import com.betfair.esa.client.cache.util.OrderMarketSnap;
import com.betfair.esa.client.protocol.ChangeMessage;
import com.betfair.esa.swagger.model.OrderMarketChange;
import java.util.ArrayList;
import java.util.EventListener;
import java.util.EventObject;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class OrderCache {
    private static final Logger LOG = LoggerFactory.getLogger(OrderCache.class);

    private final Map<String, OrderMarket> markets = new ConcurrentHashMap<>();
    private final CopyOnWriteArrayList<OrderMarketChangeListener> orderMarketChangeListeners =
            new CopyOnWriteArrayList<>();
    private final CopyOnWriteArrayList<BatchOrderMarketsChangeListener>
            batchOrderMarketsChangeListeners = new CopyOnWriteArrayList<>();

    /** Wether order markets are automatically removed on close (default is true) */
    public boolean isOrderMarketRemovedOnClose = true;

    public void onOrderChange(ChangeMessage<OrderMarketChange> changeMessage) {

        if (changeMessage.isStartOfNewSubscription()) {
            markets.clear();
        }

        if (changeMessage.getItems() != null) {
            List<OrderMarketChangeEvent> batch =
                    (batchOrderMarketsChangeListeners.size() == 0)
                            ? null
                            : new ArrayList<>(changeMessage.getItems().size());

            for (OrderMarketChange change : changeMessage.getItems()) {
                boolean isImage = Boolean.TRUE.equals(change.isFullImage());
                if (isImage) {
                    // Clear market from cache if it is being re-imaged
                    markets.remove(change.getId());
                }

                OrderMarket orderMarket = onOrderMarketChange(change);

                if (isOrderMarketRemovedOnClose && orderMarket.isClosed()) {
                    // remove on close
                    markets.remove(orderMarket.getMarketId());
                }

                if (batch != null || orderMarketChangeListeners.size() != 0) {
                    OrderMarketChangeEvent orderMarketChangeEvent =
                            new OrderMarketChangeEvent(this);
                    orderMarketChangeEvent.setChange(change);
                    orderMarketChangeEvent.setOrderMarket(orderMarket);

                    if (orderMarketChangeListeners != null) {
                        dispatchOrderMarketChanged(orderMarketChangeEvent);
                    }

                    if (batch != null) {
                        batch.add(orderMarketChangeEvent);
                    }
                }
            }

            if (batch != null) {
                dispatchBatchMarketOrderChanged(new BatchOrderMarketsChangeEvent(batch));
            }
        }
    }

    private void dispatchBatchMarketOrderChanged(
            BatchOrderMarketsChangeEvent batchOrderMarketsChangeEvent) {
        try {
            batchOrderMarketsChangeListeners.forEach(
                    l -> l.batchOrderMarketChange(batchOrderMarketsChangeEvent));
        } catch (Exception e) {
            LOG.error("Exception from batch event listener", e);
        }
    }

    private void dispatchOrderMarketChanged(OrderMarketChangeEvent orderMarketChangeEvent) {
        try {
            orderMarketChangeListeners.forEach(l -> l.orderChange(orderMarketChangeEvent));
        } catch (Exception ex) {
            LOG.error("Exception from event listener", ex);
        }
    }

    private OrderMarket onOrderMarketChange(OrderMarketChange orderMarketChange) {
        OrderMarket orderMarket =
                markets.computeIfAbsent(
                        orderMarketChange.getId(),
                        key -> new OrderMarket(this, orderMarketChange.getId()));

        orderMarket.onOrderMarketChange(orderMarketChange);
        return orderMarket;
    }

    public boolean isOrderMarketRemovedOnClose() {
        return isOrderMarketRemovedOnClose;
    }

    public void setOrderMarketRemovedOnClose(boolean orderMarketRemovedOnClose) {
        isOrderMarketRemovedOnClose = orderMarketRemovedOnClose;
    }

    public class OrderMarketChangeEvent extends EventObject {

        private OrderMarketChange orderMarketChange;
        private OrderMarket orderMarket;

        /**
         * Constructs a prototypical Event.
         *
         * @param source The object on which the Event initially occurred.
         * @throws IllegalArgumentException if source is null.
         */
        public OrderMarketChangeEvent(Object source) {
            super(source);
        }

        public OrderMarketSnap snap() {
            return orderMarket.getOrderMarketSnap();
        }

        public OrderMarketChange getChange() {
            return orderMarketChange;
        }

        public void setChange(OrderMarketChange change) {
            this.orderMarketChange = change;
        }

        public void setOrderMarket(OrderMarket orderMarket) {
            this.orderMarket = orderMarket;
        }
    }

    public class BatchOrderMarketsChangeEvent extends EventObject {

        public List<OrderMarketChangeEvent> changes;
        /**
         * Constructs a prototypical Event.
         *
         * @param source The object on which the Event initially occurred.
         * @throws IllegalArgumentException if source is null.
         */
        public BatchOrderMarketsChangeEvent(Object source) {
            super(source);
        }

        public List<OrderMarketChangeEvent> getChanges() {
            return changes;
        }

        public void setChanges(List<OrderMarketChangeEvent> changes) {
            this.changes = changes;
        }
    }

    public interface OrderMarketChangeListener extends EventListener {
        void orderChange(OrderMarketChangeEvent orderChangeEvent);
    }

    public interface BatchOrderMarketsChangeListener extends EventListener {
        void batchOrderMarketChange(BatchOrderMarketsChangeEvent batchOrderMarketsChangeEvent);
    }

    public void addOrderMarketChangeListener(OrderMarketChangeListener listener) {
        orderMarketChangeListeners.add(listener);
    }

    public void addBatchOrderMarketChangeListener(BatchOrderMarketsChangeListener listener) {
        batchOrderMarketsChangeListeners.add(listener);
    }

    public void removeBatchMarketChangeListener(BatchOrderMarketsChangeListener listener) {
        batchOrderMarketsChangeListeners.remove(listener);
    }

    public Iterable<OrderMarket> getOrderMarkets() {
        return markets.values();
    }

    public int getNumberOfOrderMarkets() {
        return markets.size();
    }
}
